libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+	Revisions            uint
+	ReadCacheStrategy    ffi.CacheStrategy
+}
+
+// Note that `FilePath` is not specified, and must always be set by the user.
+var Defaults = Config{
+	CleanCacheSize:       1024 * 1024, // 1MB
+	FreeListCacheEntries: 40_000,
+	Revisions:            100,
+	ReadCacheStrategy:    ffi.CacheAllReads,
+}
+
+func (c Config) BackendConstructor(ethdb.Database) triedb.DBOverride {
+	db, err := New(c)
+	if err != nil {
+		log.Crit("firewood: error creating database", "error", err)
+	}
+	return db
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(config Config) (*Database, error) {
+	if err := validatePath(config.ChainDir); err != nil {
+		return nil, err
+	}
+
+	// Start the logs prior to opening the database
+	logPath := filepath.Join(config.ChainDir, logFileName)
+	if err := ffi.StartLogs(&ffi.LogConfig{Path: logPath}); err != nil {
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
+		log.Warn("firewood: error starting logs", "error", err)
+	}
+
+	dbPath := filepath.Join(config.ChainDir, dbFileName)
+	fw, err := ffi.New(dbPath, &ffi.Config{
+		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		FreeListCacheEntries: config.FreeListCacheEntries,
+		Revisions:            config.Revisions,
+		ReadCacheStrategy:    config.ReadCacheStrategy,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		Firewood:    fw,
+		proposalMap: make(map[common.Hash][]*proposal),
+		proposalTree: &proposal{
+			info: &blockInfo{
+				root: common.Hash(currentRoot),
+				hashes: map[common.Hash]struct{}{
+					{}: {}, // genesis block has empty hash
+				},
+			},
+		},
+	}, nil
+}
+
+func validatePath(dir string) error {
+	if dir == "" {
+		return errors.New("chain data directory must be set")
+	}
+
+	// Check that the directory exists
+	switch info, err := os.Stat(dir); {
+	case os.IsNotExist(err):
+		log.Info("Database directory not found, creating", "path", dir)
+		if err := os.MkdirAll(dir, 0o755); err != nil {
+			return fmt.Errorf("error creating database directory: %w", err)
+		}
+		return nil
+	case err != nil:
+		return fmt.Errorf("error checking database directory: %w", err)
+	case !info.IsDir():
+		return fmt.Errorf("database directory path is not a directory: %s", dir)
+	}
+
+	return nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
```

Comment:
This can't be left to a comment here because it won't ever be done. We need to find a more robust way to enforce it, perhaps in the `triedb` package, or by returning a different scheme entirely.